src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/AsyncLazyWithArg`2.cs

```diff
@@ -0,0 +1,77 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Roslyn.Utilities;
+
+/// <summary>
+/// Represents an AsyncLazy instance holding compute functions and additional 
+/// argument data to pass to those functions
+/// </summary>
+/// <typeparam name="T">Return type of compute functions</typeparam>
+/// <typeparam name="TArg">Argument type to compute functions</typeparam>
+internal sealed class AsyncLazyWithArg<T, TArg> : AsyncLazy<T>
+{
+    private PooledDelegates.Releaser _asyncComputeFunctionReleaser;
+    private PooledDelegates.Releaser _syncComputeFunctionReleaser;
+    private bool _isAsyncComputeFunctionReleaserDisposed;
+    private bool _isSyncComputeFunctionReleaserDisposed;
+
+    public AsyncLazyWithArg(Func<TArg, CancellationToken, Task<T>> asynchronousComputeFunction, TArg arg)
+        : this(asynchronousComputeFunction, synchronousComputeFunction: null, arg)
+    {
+    }
+
+    public AsyncLazyWithArg(Func<TArg, CancellationToken, Task<T>> asynchronousComputeFunction, Func<TArg, CancellationToken, T>? synchronousComputeFunction, TArg arg)
```

Comment:
>AsyncLazyWithArg [](http://example.com/codeflow?start=11&length=16)

Looking at this I see it's ugly. Let me rework...

Reply:
Non-copyable structs are making cleanup difficult for me. I'm just going to leave as is as I don't really want to devote any more time as I'm not certain this is the approach we want anyways.